Unary unary client aio call timeout

rmariano opened the pull request:
Tested with the new unit (former e2e) tests:
```
$ python src/python/grpcio_tests/tests_aio/unit/channel_test.py

test_async_context (__main__.TestChannel) ... ok
test_unary_call_survives_timeout (__main__.TestChannel) ... ok
test_unary_call_times_out (__main__.TestChannel) ... ok
test_unary_unary (__main__.TestChannel) ... ok

----------------------------------------------------------------------
Ran 4 tests in 0.533s

OK
```

Fixes https://github.com/grpc/grpc/issues/19871

pfreixes commented on src/python/grpcio_tests/tests_aio/unit/sync_server.py:
```diff
@@ -30,6 +30,10 @@ class TestServiceServicer(test_pb2_grpc.TestServiceServicer):
     def UnaryCall(self, request, context):
         return messages_pb2.SimpleResponse()
 
+    def EmptyCall(self, request, context):
```
better `TimedoutCall`?

rmariano replied:
mm the name has to respect the one on the base class. So for this, I would have to create a new protobuf (or update the current one, which I don't think would be allowed), to add this method.

pfreixes replied:
True ... Im starting to think that for the sake of having everything decoupled we would need to start using our own protobuf files for the asynchronous tests, so having full freedom for creating new methods and not breaking things. I'm gonna start a discussion in the channel regarding this.

pfreixes commented on src/python/grpcio_tests/tests_aio/unit/sync_server.py:
```diff
@@ -30,6 +30,10 @@ class TestServiceServicer(test_pb2_grpc.TestServiceServicer):
         return messages_pb2.SimpleResponse()
 
+    def EmptyCall(self, request, context):
+        sleep(0.2)
```
Having the feeling that this would need to be a constant, used by the server and the client.

pfreixes commented on src/python/grpcio_tests/tests_aio/unit/channel_test.py:
```diff
@@ -52,6 +52,34 @@ def test_unary_unary(self):
+                request_serializer=messages_pb2.SimpleRequest.SerializeToString,
+                response_deserializer=messages_pb2.SimpleResponse.FromString,
+            )
+            await empty_call_with_sleep(messages_pb2.SimpleRequest(), timeout=0.1)
```
a timed out call does not end up with an exception? If so we should implement the same behavior as the synchronous client.

pfreixes commented on src/python/grpcio_tests/tests_aio/unit/channel_test.py:
```diff
@@ -52,6 +52,34 @@ def test_unary_unary(self):
 
         self.loop.run_until_complete(coro())
 
+    def test_unary_call_survives_timeout(self):
```
IMO this test would not be necessary.

pfreixes commented on src/python/grpcio/grpc/experimental/aio/_channel.py:
```diff
@@ -50,10 +48,11 @@ def __init__(self, channel, method, request_serializer,
+        response = await self._channel.unary_unary(self._method, serialized_request, timeout)
 
-        return _common.deserialize(response, self._response_deserializer)
+        return _common.deserialize(response or b"", self._response_deserializer)
```
`response or b""` having the feeling that you have been forced to do so since a timed out call follows the same path as the happy path. Having the feeling that a timed out call should raise an exception, and if so this would not be necessary.

Whats the current behavior when a timeout is raised in the synchronous API?

rmariano replied:
In the synchronous API, the server calls ``.cancel()`` and returns a message with the error, stating that the deadline has been exceeded. In that case, shouldn't then this be implemented on the server, once we have the asynchronous one available? Maybe a TODO?

pfreixes replied:
Correct me if I'm wrong but theoretically, we are using a server with full batteries included - the synchronous one, so theoretically we should be able to catch this error and return back to the user the proper explicit message telling that the call finished with a timeout.

Once we start using the Asynchronous server, the test should remain green, otherwise would mean that something wrong is implemented in the server.

Also, I'm wondering if a deadline can be only be raised by the Server. Does the client have a mechanism for telling explicitly that the timeout was reached? Considering that the network or the server can fail which will result in not telling the client that the deadline was reached.  We would need to clarify this, and if see if both use cases share the same path.

pfreixes replied:
Regarding 

> Also, I'm wondering if a deadline can be only be raised by the Server

I didn't test it but this is what is being said in the gRPC documentation [1]

> In gRPC, both the client and server make their own independent and local determination about whether the remote procedure call (RPC) was successful. This means their conclusions may not match! An RPC that finished successfully on the server side can fail on the client side. For example, the server can send the response, but the reply can arrive at the client after their deadline has expired. The client will already have terminated with the status error DEADLINE_EXCEEDED. This should be checked for and managed at the application level.

[1] https://grpc.io/blog/deadlines/

pfreixes commented on src/python/grpcio/grpc/experimental/aio/_channel.py:
```diff
@@ -50,10 +48,11 @@ def __init__(self, channel, method, request_serializer,
-        response = await self._channel.unary_unary(
-            self._method, _common.serialize(request, self._request_serializer))
+        serialized_request = _common.serialize(request, self._request_serializer)
+        timeout = None if timeout is None else time.time() + timeout
```
theoretically, we should use `loop.time()` function.

pfreixes commented on src/python/grpcio/grpc/_cython/_cygrpc/aio/iomgr/iomgr.pyx.pxi:
```diff
@@ -33,7 +33,7 @@ cdef grpc_error* asyncio_socket_init(
 
 cdef void asyncio_socket_destroy(grpc_custom_socket* grpc_socket) with gil:
-    Py_DECREF(<_AsyncioSocket>socket.impl)
+    Py_DECREF(<_AsyncioSocket> grpc_socket.impl)
```
good catch.

argaen commented on src/python/grpcio/grpc/_cython/_cygrpc/aio/__init__.py:
```diff
@@ -0,0 +1,13 @@
+# Copyright 2019 gRPC authors.
```
I think in the original PR this and the `iomgr/__init__.py` files had to be removed

argaen commented on src/python/grpcio/grpc/_cython/_cygrpc/aio/iomgr/iomgr.pyx.pxi:
```diff
@@ -51,13 +51,13 @@ cdef void asyncio_socket_close(
         grpc_custom_socket* grpc_socket,
         grpc_custom_close_callback close_cb) with gil:
     socket = (<_AsyncioSocket>grpc_socket.impl)
-    if socket.is_connected():
```
why don't we need this check anymore?

argaen replied:
ah, I see the check is moved inside the socket class

argaen commented on src/python/grpcio_tests/tests_aio/unit/test_base.py:
```diff
@@ -32,17 +32,19 @@ def _get_free_loopback_tcp_port():
+        cls._server = sync_server.Server(cls._target)
+        cls._server.start()
+        cls._loop = asyncio.new_event_loop()
+        asyncio.set_event_loop(cls._loop)
```
isn't this removing the isolation between tests? Are we happy about this? No strong opinion, just interested in knowing the rationale behind this

rmariano replied:
The problem I was facing (7b7c751b3b63fe91df170d92bdcca4d360462a0c) was that if any of the test crashes, it wasn't propertly freeing up resources, so the next test that tried to run will fail trying to bind the host&port, and at the end of the tests, the server was still left running, leaking file descriptors, and still using the port.

pfreixes replied:
In an ideal world, I would prefer having each test case using its own isolated fixture rather than using a global one. Otherwise, we have the risk of leaving the server in a bad state for the following test.

Reading the `tearDown` documentation it says that 

>This is called even if the test method raised an exception, so the implementation in subclasses may need to be particularly careful about checking internal state
  
So theoretically every time that a test finishes with an error the `tearDown` method would need to be executed yes or yes ... If this is not happening I would try to find the reason why the server is not being released after a test failure.

pfreixes replied:
Could be this issue related to this [1]? Im still skeptical on that change, tests pass on my computer so we would need proof - stacktrace ? - that tests are broken without this.

[1] https://github.com/grpc/grpc/commit/8a80f8792a72b13ea60ded0a8385ec08fb91c863

pfreixes commented on src/python/grpcio_tests/tests_aio/unit/channel_test.py:
```diff
@@ -52,6 +54,21 @@ def test_unary_unary(self):
+            )
+            timeout = self.CALL_DELAY / 2
+            await empty_call_with_sleep(messages_pb2.SimpleRequest(), timeout=timeout)
+            await channel.close()
```
I would encourage using the async context for not having to call explicitly every time the `close` method once the channel is not longer used.

pfreixes commented on src/python/grpcio_tests/tests_aio/unit/channel_test.py:
```diff
@@ -52,6 +54,21 @@ def test_unary_unary(self):
+            await empty_call_with_sleep(messages_pb2.SimpleRequest(), timeout=timeout)
+            await channel.close()
+
+        with self.assertRaisesRegex(grpc.RpcError, r".*Deadline Exceeded.*"):
```
I would move this within the `coro` function.

Another question is, what does prevent us of using the same test pattern used in the synchronous client [1] that checks explicitly the `grpc.StatusCode.DEADLINE_EXCEEDED` that seems more readable than using a reg expression?


[1] https://github.com/grpc/grpc/blob/master/src/python/grpcio_tests/tests/unit/_rpc_test.py#L600

rmariano replied:
I am moving this check to the ``_coro()``
About the second point, here I'd like your opinion: The reason for this check is because the other implementation does
```python
raise _Rendezvous(code, None, None, ... )
```
Where ``_Rendezvous`` is a custom private object that inherics from ``(grpc.RpcError, grpc.Future, grpc.Call)``.
Does it make sense to create an object like this one here?
Would you agree on creating another custom exception with the same methods? (``.code(), ...`` and the rest)
Also keep in mind that these objects are defined in Python, so we cannot extend them from cython code.

rmariano replied:
This is what I have in mind
```python
class _RpcError(grpc.RpcError, grpc.Future, grpc.Call):  # pylint: disable=too-many-ancestors

    def __init__(self, initial_metadata, code, details, trailing_metadata):
        self._initial_metadata = initial_metadata
        self._code = code
        self._details = details
        self._trailing_metadata = trailing_metadata

    def initial_metadata(self):
        return self._initial_metadata

    def code(self):
        return self._code

    def details(self):
        return self._details

    def trailing_metadata(self):
        return self._trailing_metadata
```
The only issue, I'm working on right now is that importing ``grpc`` gives an error from the cython module (it fails at the definition of ``grpc.RpcError``).

I will continue with this line of working, unless you have something else in mind.

pfreixes commented on src/python/grpcio_tests/tests_aio/unit/test_base.py:
```diff
@@ -32,17 +32,21 @@ def _get_free_loopback_tcp_port():
-        self._server.start()
-        self._loop = asyncio.new_event_loop()
-        asyncio.set_event_loop(self._loop)
+    CALL_DELAY = sync_server.TestServiceServicer.CALL_DELAY
```
Im not really fan of this, can we use the constant declared into the `sync_server` module?

rmariano commented:
Hi @pfreixes I have addressed your comments in 87c0f759691b74dd3e326efc3bcdc0d0b616bc5c
About the tests done per class: I tested again, and verified that indeed they don't work with the former approach, when there are more than one test doing RPC calls to the channel, so I believe we need this setup for now.

rmariano commented:
Tested with the latest changes
```
$ GRPC_ENABLE_FORK_SUPPORT=0 python ./src/python/grpcio_tests/setup.py test_aio
running test_aio
Running       _sanity._sanity_test.AioSanityTest.testTestsJsonUpToDate
Testing gRPC Python...
SUCCESS       _sanity._sanity_test.AioSanityTest.testTestsJsonUpToDate
Running       unit.channel_test.TestChannel.test_async_context
SUCCESS       unit.channel_test.TestChannel.test_async_context
Running       unit.channel_test.TestChannel.test_unary_call_times_out
SUCCESS       unit.channel_test.TestChannel.test_unary_call_times_out
Running       unit.channel_test.TestChannel.test_unary_unary
SUCCESS       unit.channel_test.TestChannel.test_unary_unary
Running       unit.init_test.TestInitGrpcAio.test_aio_supports_fork_and_grpc_aio_usable
SUCCESS       unit.init_test.TestInitGrpcAio.test_aio_supports_fork_and_grpc_aio_usable
Running       unit.init_test.TestInitGrpcAio.test_aio_supports_fork_and_grpc_blocking_usable
SUCCESS       unit.init_test.TestInitGrpcAio.test_aio_supports_fork_and_grpc_blocking_usable
Running       unit.init_test.TestInsecureChannel.test_insecure_channel
SUCCESS       unit.init_test.TestInsecureChannel.test_insecure_channel
7 tests finished:
	7 successful
	0 unsuccessful
	0 skipped
	0 expected failures
	0 unexpected successes
Interrupted Tests:
	[]

Errors/Failures:

Unexpected successes: []
```

rmariano commented:
@pfreixes I have applied your latest feedback

rmariano commented:
This is deprecated. Closing in favour of https://github.com/grpc/grpc/pull/20277
